src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -1099,6 +1243,99 @@ void PersistNode(TrieNode tn, Hash256? address2, TreePath path)
+            TreePath path = TreePath.Empty;
+            commitSet.Root?.CallRecursively(TopLevelPersist, address, ref path, GetTrieStore(null), true, _logger, maxPathLength: parallelBoundaryPathLength);
+
+            // The amount of change in the subtrees are not balanced at all. So their writes ares buffered here
+            // which get disposed in parallel instead of being disposed in `PersistNodeStartingFrom`.
+            // This unfortunately is not atomic
+            // However, anything that we are trying to persist here should still be in dirty cache.
+            // So parallel read should go there first instead of to the database for these dataset,
+            // so it should be fine for these to be non atomic.
+            BlockingCollection<INodeStorage.WriteBatch> disposeQueue = new BlockingCollection<INodeStorage.WriteBatch>(4);
+
+            Task disposeTask = Task.WhenAll(Enumerable.Range(0, Environment.ProcessorCount).Select((_) => Task.Run(() =>
+            {
+                while (disposeQueue.TryTake(out INodeStorage.WriteBatch disposable, Timeout.Infinite))
+                {
+                    disposable.Dispose();
+                }
+            })).ToArray());
+
+            Task.WaitAll(parallelStartNodes.Select((entry) => Task.Run(() =>
+            {
+                (TrieNode trieNode, Hash256? address2, TreePath path2) = entry;
+                PersistNodeStartingFrom(trieNode, address2, path2, commitSet, persistedNodeRecorder, writeFlags, disposeQueue);
+            })).ToArray());
+
+            disposeQueue.CompleteAdding();
+            disposeTask.Wait();
+
+            // Dispose top level last in case something goes wrong, at least the root wont be stored
+            topLevelWriteBatch.Dispose();
+
+            stopwatch.Stop();
+            Metrics.SnapshotPersistenceTime = stopwatch.ElapsedMilliseconds;
+
+            if (_logger.IsDebug) _logger.Debug($"Persisted trie from {commitSet.Root} at {commitSet.BlockNumber} in {stopwatch.ElapsedMilliseconds}ms (cache memory {MemoryUsedByDirtyCache})");
+
+            LastPersistedBlockNumber = commitSet.BlockNumber;
+        }
+
+        private void PersistNodeStartingFrom(TrieNode tn, Hash256 address2, TreePath path, BlockCommitSet commitSet,
+            Action<TreePath, Hash256?, TrieNode>? persistedNodeRecorder,
+            WriteFlags writeFlags, BlockingCollection<INodeStorage.WriteBatch> disposeQueue)
+        {
+            long persistedNodeCount = 0;
+            INodeStorage.WriteBatch writeBatch = _nodeStorage.StartWriteBatch();
+
+            void DoPersist(TrieNode node, Hash256? address3, TreePath path2)
+            {
+                persistedNodeRecorder?.Invoke(path2, address3, node);
+                PersistNode(address3, path2, node, commitSet.BlockNumber, writeFlags, writeBatch);
+
+                persistedNodeCount++;
+                if (persistedNodeCount % 512 == 0)
```

Comment:
Yes, this is what I think the best value for 4GB cache. Hard to see any difference though.

Reply:
Why small batches are better than large batches?
Each batch needs to be sorted - it could be done in memory.
Then it kind of needs to do a linear scan for insertions.
It kind of makes sense that a smaller amount of larger batches would be faster?

Reply:
Its not that simple. 

What we are trying to do here (ignoring WAL), is to dump data to the memtable which is a skiplist.
Because of how atomicity works with rocksdb, a number of batches group together, and they operate on the skiplist concurrently, but all of them only returns after the slowest batch finish operating on the memtable. Because of that, 
we want to have the memtables to be of similar size. Hence, the batch size limit. This is unless unordered writes options is turned on, but we lose the atomicity of the batch. (a concurrent read might read one item from the batch, but not the other while the batch is being written).

The sorted order is likely due to the fact that the skiplist implementation have a fast path where it can skip traversing a good chunk of the skiplist if the written key is after the last written key. The question is, does this occur a lot when multiple thread writes to it? Does not seems like so. 

Yes, if one of the slow batch in a batch group is sorted, then it is a noticable improvement. But then later I just dump the batches to a blocking collection and its so much faster in any case.